### ISSUE TITLE
with_file_input_stream doesn't close file

### ISSUE THREAD

**Issue by xemul** [scylladb/seastar#1298]
as per fe8fcd6d2fded7c94bd3e8fc3e489e3d9ba060ab
```c++
auto with_file_input_stream(const std::filesystem::path& path, Func func, file_open_options file_opts = {}, file_input_stream_options input_stream_opts = {}) {
    static_assert(std::is_nothrow_move_constructible_v<Func>);
    return open_file_dma(path.native(), open_flags::ro, std::move(file_opts)).then(
            [func = std::move(func), input_stream_opts = std::move(input_stream_opts)] (file f) mutable {
        return do_with(make_file_input_stream(std::move(f), std::move(input_stream_opts)),
                [func = std::move(func)] (input_stream<char>& in) mutable {
            return futurize_invoke(std::move(func), in).finally([&in] {
                return in.close();
            });
        });
    }); 
}
```

the file is move-d into input stream, but closing the stream doesn't close the file, respectively file is not closed (however, file_impl-s are supposed to close the underlying descriptor in destructor)

cc @bhalevy

**Comment by bhalevy** [comment]
Hmm, it seems like file_data_source_impl::close should close its _file member.
Is there a reproducer for this issue?
Best if we had a unit test for it.

**Comment by xemul** [comment]
> Hmm, it seems like file_data_source_impl::close should close its _file member.

Of course, the stream itself doesn't have file on board, the issue is rather about data source

> Is there a reproducer for this issue?

Hmm, I don't think there is one. But what kind of "reproducer" should it be? Is there any landmine we can step on if leaving a file non-closed?

**Comment by bhalevy** [comment]
> > Hmm, it seems like file_data_source_impl::close should close its _file member.
> 
> Of course, the stream itself doesn't have file on board, the issue is rather about data source
> 
> > Is there a reproducer for this issue?
> 
> Hmm, I don't think there is one. But what kind of "reproducer" should it be? Is there any landmine we can step on if leaving a file non-closed?

The critical part is cancelling outstanding reads and waiting for them to resolve before destroying the file.
Otherwise, we risk use after free.

**Comment by niekbouman** [comment]
> [bhalevy]: Hmm, it seems like file_data_source_impl::close should close its _file member.

Could it be that you were confusing `file_data_sink_impl::close` and `file_data_source_impl::close`?

`file_data_sink_impl` does close the file in its `close` virtual method:
https://github.com/scylladb/seastar/blob/master/src/core/fstream.cc#L492

but it seems that `file_data_source_impl` does not:
https://github.com/scylladb/seastar/blob/master/src/core/fstream.cc#L278

**Comment by bhalevy** [comment]
> > [bhalevy]: Hmm, it seems like file_data_source_impl::close should close its _file member.
> 
> Could it be that you were confusing `file_data_sink_impl::close` and `file_data_source_impl::close`?
> 
> `file_data_sink_impl` does close the file in its `close` virtual method: https://github.com/scylladb/seastar/blob/master/src/core/fstream.cc#L492
> 
> but it seems that `file_data_source_impl` does not: https://github.com/scylladb/seastar/blob/master/src/core/fstream.cc#L278

I said that:
> it seems like file_data_source_impl::close **should** close its _file member.

I think we're in agreement.

**Comment by raphaelsc** [comment]
> > > [bhalevy]: Hmm, it seems like file_data_source_impl::close should close its _file member.
> > 
> > 
> > Could it be that you were confusing `file_data_sink_impl::close` and `file_data_source_impl::close`?
> > `file_data_sink_impl` does close the file in its `close` virtual method: https://github.com/scylladb/seastar/blob/master/src/core/fstream.cc#L492
> > but it seems that `file_data_source_impl` does not: https://github.com/scylladb/seastar/blob/master/src/core/fstream.cc#L278
> 
> I said that:
> 
> > it seems like file_data_source_impl::close **should** close its _file member.
> 
> I think we're in agreement.

it's the correct thing to do, but its absence shouldn't be causing any issue today right? we wait for outstanding requests. so close in dtor shouldn't be blocking reactor.

**Comment by bhalevy** [comment]
> > > > [bhalevy]: Hmm, it seems like file_data_source_impl::close should close its _file member.
> > > 
> > > 
> > > Could it be that you were confusing `file_data_sink_impl::close` and `file_data_source_impl::close`?
> > > `file_data_sink_impl` does close the file in its `close` virtual method: https://github.com/scylladb/seastar/blob/master/src/core/fstream.cc#L492
> > > but it seems that `file_data_source_impl` does not: https://github.com/scylladb/seastar/blob/master/src/core/fstream.cc#L278
> > 
> > 
> > I said that:
> > > it seems like file_data_source_impl::close **should** close its _file member.
> > 
> > 
> > I think we're in agreement.
> 
> it's the correct thing to do, but its absence shouldn't be causing any issue today right? we wait for outstanding requests. so close in dtor shouldn't be blocking reactor.

Trying to add said close in `file_data_source_impl::close` is problematic since there are tests and possibly application that already rely on it to not close the file.
The reason is that `seastar::file` contains a shared pointer of the file_impl and it is allowed (and quite typical) for the caller of `make_file_input_stream` to hold a copy of the seastar::file and use it.
Closing the file in `file_data_source_impl::close` closes the file handle and this also closes the copy of that file, since we're not reference counting the file close():s (if we would, then we could close the file_impl on last close()).

Bottom line is that we can fix the file close only in `with_file_input_stream` since the file object there is local to the function.
